README.md
# Haskellator

Calculator written in Haskell with the Brick TUI library.
The vision of Haskellator is a quick-to-use terminal calculator supporting scientific functions and saveable variables.
I really like applications with a very minimalistic user interface that's intuitive to use.

## FEATURES UNDER CONSTRUCTION

1. The TUI (Brick)
2. The Haskellator language (supporting arithmetic expressions and variable assignment)

# Contribute

Unfortunately, the project is closed to contributions at this time.
I plan on accepting contributions once the application has made it out of initial development.


CHANGELOG.md
# Revision history for helloWorld

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


Eval.hs
module Eval where 
-- below we are going to add fucntions for handling parsing/evaluating from user input
-- these are instructions to handle with parsing

type Id = String 
type N = Double 

data Instr = Assign Id Expr
    | AExpr Expr 
    deriving Show 

data Expr = Const N
    |Add Expr Expr | Sub Expr Expr
    |Mul Expr Expr | Div Expr Expr
    |Expo Expr Expr -- | Mod Expr Expr
    |Sin Expr | Cos Expr
    deriving Show 

-- type of environment for variables to be assigned in
--type Env = [(Id,N)]

-- to find values that are associated with given variables
--lookUp :: String -> Env -> Double
--lookUp a [] = error "there is no pair associated with this String"
--lookUp a ((x,y):zs) = if (a==x) then y else lookUp a zs

--this function adds a var to a env list 
--addVar :: (String, Double) -> Env -> Env 
--addVar x [] = [x]
--addVar x env = x : env 

-- when calling update we will need to call remove to replace the env 
--removeVar :: String -> Env -> Env 
--removeVar a [] = []
--removeVar a ((x,y):zs) = if a == x then zs else (x,y) : removeVar a zs

-- update will set value of given parameter to new value and keep everything else the same
--update :: String -> Double -> Env -> Env 
--update s d [] = [(s,d)]
--update s d env = addVar (s,d) (removeVar s env)

eval :: Instr -> Either Double (Id, N)
eval (Assign x xx) = Right (x, evalExp xx)
eval (AExpr x) = Left (evalExp x)

--handling evaluating our exp 
evalExp :: Expr -> Double
evalExp (Const x) = x
evalExp (Add x xx) = (evalExp x) + (evalExp xx)
evalExp (Sub x xx) = (evalExp x) - (evalExp xx)
evalExp (Mul x xx) = (evalExp x) * (evalExp xx) 
evalExp (Div x xx) = (evalExp x) / (evalExp xx)
evalExp (Expo x xx) = (evalExp x) ** (evalExp xx)
evalExp (Cos x) = cos(evalExp x)
evalExp (Sin x) = sin(evalExp x)
--come back to mod because it does not really have a double value way could cast using tointegral but who knows 
--evA env (Mod x xx) = (evA env x) % (evA env xx) 


Lang.hs
module Lang (exec) where

type Id  = String
type Val = Double

type Env = [(Id, Val)]

data Instr = IAssign Id Expr
           | IExpr Expr
           deriving (Show, Eq)

data Expr = Const Val
          | Var Id
          | Add Expr Expr
          | Sub Expr Expr
          | Mul Expr Expr
          | Div Expr Expr
          deriving (Show, Eq)

data Token = TId Id | TConst Val
           | TEq | TAdd | TSub | TMul
           | TDiv | TLPar | TRPar
           | TExpr Expr
           | TAssign Id Expr
           deriving (Show, Eq)

-- Test environment
testEnv :: Env
testEnv = [("pi", 3.1415), ("x", 2.2)]

envLookup :: Id -> Env -> Val
envLookup _ [] = error "Variable not in environment."
envLookup k ((id, val):xs) | k == id   = val
                           | otherwise = envLookup k xs

envAdd :: (Id, Val) -> Env -> Env
envAdd var e = var : e

envRemove :: Id -> Env -> Env
envRemove _ [] = []
envRemove k (v@(id, _):xs) | k == id   =     envRemove k xs
                           | otherwise = v : envRemove k xs

envUpdate :: (Id, Val) -> Env -> Env
envUpdate x e = envAdd x $ envRemove (fst x) e

evalInstr :: Instr -> Env -> Either Val Env
evalInstr (IAssign id expr) env = Right (envUpdate (id, evalExpr expr env) env)
evalInstr (IExpr expr) env = Left (evalExpr expr env)

evalExpr :: Expr -> Env -> Val
evalExpr (Const v) e = v
evalExpr (Var id) e = envLookup id e
evalExpr (Add e1 e2) e = evalExpr e1 e + (evalExpr e2 e)
evalExpr (Sub e1 e2) e = evalExpr e1 e - (evalExpr e2 e)
evalExpr (Mul e1 e2) e = evalExpr e1 e * (evalExpr e2 e)
evalExpr (Div e1 e2) e = evalExpr e1 e / (evalExpr e2 e)



isCSym :: String -> Bool
isCSym [] = True
isCSym (x:xs) | x `elem` ['0' .. '9'] = isCSym xs
              | x == '.' = isCSym' xs
              | otherwise = False
                  where isCSym' [] = True
                        isCSym' (y:ys) = if y `elem` ['0' .. '9'] then isCSym' ys else False

isVSym :: String -> Bool
isVSym [] = False
isVSym (x:xs) = if x `elem` ['a' .. 'z'] then isVSym' xs else False

isVSym' :: String -> Bool
isVSym' [] = True
isVSym' (x:xs) = if x `elem` ['a' .. 'z'] ++ ['1' .. '9'] then isVSym' xs else False

classify :: String -> Token
classify s | s == "+" = TAdd
           | s == "-" = TSub
           | s == "*" = TMul
           | s == "/" = TDiv
           | s == "=" = TEq
           | s == "(" = TLPar
           | s == ")" = TRPar
           | isCSym s = TConst (read s :: Val)
           | isVSym s = TId s

prep :: String -> String
prep [] = []
prep ('+':xs) = " + " ++ prep xs
prep ('-':xs) = " - " ++ prep xs
prep ('*':xs) = " * " ++ prep xs
prep ('/':xs) = " / " ++ prep xs
prep ('=':xs) = " = " ++ prep xs
prep ('(':xs) = " ( " ++ prep xs
prep (')':xs) = " ) " ++ prep xs
prep (x:xs) = x : prep xs 

tokenize :: String -> [Token]
tokenize s = map classify $ words $ prep s


-- Parse time baby
parse :: [Token] -> [Token] -> [Token]
parse ((TConst c):stack) input = parse (TExpr (Const c) : stack) input

-- Var
parse ((TId id):stack) input = parse (TExpr (Var id) : stack) input

-- Add, Sub, Mul, Div
parse ((TExpr e2):TDiv:(TExpr e1):stack) input = parse (TExpr (Div e1 e2):stack) input
parse ((TExpr e2):TMul:(TExpr e1):stack) input = parse (TExpr (Mul e1 e2):stack) input
parse ((TExpr e2):TAdd:(TExpr e1):stack) input = parse (TExpr (Add e1 e2):stack) input
parse ((TExpr e2):TSub:(TExpr e1):stack) input = parse (TExpr (Sub e1 e2):stack) input

-- Assignment
parse ((TExpr e):TEq:(TExpr (Var id)):stack) input = parse (TAssign id e : stack) input

-- Parens
parse (TRPar : expr@(TExpr e) : TLPar : stack) input = parse (expr:stack) input

-- Defaults
parse stack (i:input) = parse (i:stack) input
parse stack [] = stack



tti :: Env -> [Token] -> Either Val Env
tti env [(TExpr e)]      = evalInstr (IExpr e) env
tti env [(TAssign id e)] = evalInstr (IAssign id e) env
tti _ _                  = error "tti error: too many instructions?" 

exec :: String -> Env -> Either Val Env
exec s e = tti e $ parse [] $ tokenize s


Parser.hs
module Parser where

import Eval
import Prepro


-- data BOps = AddOp | SubOp | MulOp | DivOp | ExpOp
--        
--           deriving (Show,Enum,Eq,Ord)

-- data UOps = CosOp | SinOp 
--           deriving Show 

-- data Token = Tcon N | TId Id
--            | UOp UOps | BOp BOps 
--            | LPar | RPar | Err 
--            | PE Expr  | PAss Id Expr 
--            deriving Show  

-- here we are goin to do all the parsing in this file 
parse :: [Token] -> [Token] -> [Token]
parse (Err : stack) input = error "somthing wacky on the stacky yo!"
parse (Tcon v : stack) input = parse (PE (Const v): stack) input

parse s@(PE ee : BOp op1 : PE e : stack) (BOp op2 : input) | op1 < op2 = parse (BOp op2 : s) input -- taking care of presidence in an Expr 

parse (RPar: PE e : LPar : stack)        input = parse (PE e : stack) input -- taking care of parens we just get rid of them 

parse (PE ee : BOp AddOp : PE e : stack) input = parse (PE (Add e ee) : stack) input -- covers add
parse (PE ee : BOp SubOp : PE e : stack) input = parse (PE (Sub e ee) : stack) input -- covers Sub
parse (PE ee : BOp MulOp : PE e : stack) input = parse (PE (Mul e ee) : stack) input -- covers mul
parse (PE ee : BOp DivOp : PE e : stack) input = parse (PE (Add e ee) : stack) input -- covers div 
parse (PE ee : BOp ExpOp : PE e : stack) input = parse (PE (Expo e ee) : stack) input

parse (PE e : UOp CosOp : stack)         input = parse (PE(Cos e):stack) input -- taking care of cos
parse (PE e : UOp SinOp : stack)         input = parse (PE(Sin e):stack) input -- taking care of sin

parse (PE ee: OpAss : TId e : stack)     input = parse (PAss e ee : stack ) input -- taking care of an assignmnet 

tokToIn :: [Token] -> Instr
tokToIn [] = error "Nothing to parse to instruction"
tokToIn [(PE e)] = AExpr e 
tokToIn [(PAss id e)] = Assign id e 
tokToIn _ = error "Parse Error"

exec :: String -> Either Double (Id, N)
exec s = eval $ tokToIn $ parse [] $ lexer s

Prepro.hs
module Prepro where

import Eval
import Data.Char


data BOps = AddOp | SubOp | MulOp | DivOp | ExpOp
          deriving (Show,Enum,Eq,Ord)

data UOps = CosOp | SinOp 
          deriving Show 

data Token = Tcon N | TId Id
           | UOp UOps | BOp BOps 
           | LPar | RPar | Err 
           | OpAss
           | PE Expr | PAss Id Expr 
           deriving Show 

classify :: String -> Token
classify "+" = BOp AddOp
classify "-" = BOp SubOp
classify "*" = BOp MulOp
classify "=" = OpAss
classify "^" = BOp ExpOp
--classify "==" = BOp EqlOp
classify "cos" = UOp CosOp 
classify "sin" = UOp SinOp
classify "(" = LPar
classify ")" = RPar
classify x | isVSym x = TId x
           | isCSym x = Tcon (read x :: Double)
           | otherwise = Err
--classify _ = Err 

preproc :: String -> String 
preproc [] = []
preproc ('(':xs) = " ( " ++ preproc xs
preproc (')':xs) = " ) " ++ preproc xs
preproc ('+':xs) = " + " ++ preproc xs
preproc ('-':xs) = " - " ++ preproc xs
preproc ('*':xs) = " * " ++ preproc xs
preproc ('^':xs) = " ^ " ++ preproc xs
preproc ('=':'=':xs) = " == " ++ preproc xs
preproc ('=':xs) = " = " ++ preproc xs
preproc ('c':'o':'s':xs) = " cos " ++ preproc xs
preproc ('s':'i':'n':xs) = " sin " ++ preproc xs
preproc (x:xs) = x : preproc xs 

lexer :: String -> [Token]
lexer s = map classify $ words $ preproc s

-- checking for correct number value 123 decimal followed by 12312 no more decimals 
isCSym:: String -> Bool 
isCSym"" = False
isCSym(x:xs) = isDigit x && q1 xs 
    where q1 "" = True 
    -- check the state after a decimal is invloved 
          q1 ('.':ys) = q2 ys
            where q2 "" = True
                  q2 (z:zs) = isDigit z && q2 zs
    -- if before a decimal then keep checking isDigit
          q1 (y:ys) = isDigit y && q1 ys 

isVSym :: String -> Bool 
isVSym "" = False
isVSym (x:xs) = (isUpper x || isLower x) && q1 xs 
    where q1 "" = True
          q1 (y:ys) = (isUpper y || isLower y ) && q1 ys 


Haskellator.hs
module Main where

import Lang

import Brick
import Brick.AttrMap
import Brick.Util
import Brick.Widgets.Border
import Brick.Widgets.Center
import Brick.Widgets.Dialog
import Brick.Widgets.Table
import Graphics.Vty.Input.Events

import Cursor.Simple.List.NonEmpty
import qualified Data.List.NonEmpty as DNE
import Graphics.Vty
import System.Exit
import Data.Char

    -- Main
main :: IO ()
main = do
    initState <- buildInitState
    endState <- defaultMain app initState
    print endState

-- My own state for the program
data MyState = MyState {
    evalString :: String,
    env        :: [(String, Double)]
    }
    deriving (Show, Eq)

-- Resource name is simply typedef of String
type ResourceName = String

-- Creating the app
app :: App MyState e ResourceName
app = App {
    appDraw = drawUI,
    appChooseCursor = neverShowCursor,
    appHandleEvent = handleEvent,
    appStartEvent = pure,
    appAttrMap = const $ attrMap mempty [(attrName "selected", fg red)]
    }

-- Takes a state and returns a list of widgets of
-- the appropriate resource name
drawUI :: MyState -> [Widget ResourceName]
drawUI s = [(hCenter $ str $ "Haskellator v0.1") <=> (padTop (Pad 2) $ hCenter $ border $ vBox $ map str $ envConcat $ env s)
    ,center $ border $ str $ evalString s
    ]

-- Helper function to compress the MyState env to a widget compatible list
envConcat :: [(String, Double)] -> [String]
envConcat [] = []
envConcat (x:xs) = (fst x ++ " = " ++ (show $ snd x)) : envConcat xs

-- buildInitState: bui
buildInitState :: IO MyState
buildInitState = pure MyState { evalString = "", env = [] }

-- Handling vty events
handleEvent :: MyState -> BrickEvent n e -> EventM n (Next MyState)
handleEvent s e =
    case e of
        VtyEvent vtye ->
            case vtye of
            EvKey (KChar 'q') [] -> halt s
            EvKey KBS [] -> continue s'
                where s' = if null $ evalString s then s
                           else MyState { evalString = init $ evalString s, env = env s }
            EvKey (KChar c) [] -> continue MyState {evalString = evalString s ++ [c], env = env s }
            EvKey KEnter [] -> continue s'
                where s' = case exec (evalString s) (env s) of
                           (Left val)     -> MyState {evalString = show val, env = env s}
                           (Right newEnv) -> MyState {evalString = "", env = newEnv}
            _ -> continue s
        _ -> continue s


Parse.hs
module Parse where 

import Data.Char

type Id = String 
type N = Double 

data Instr = Assign Id Expr
           | AExpr Expr 
           deriving Show 

data Expr = Var Id | Const N
          |Add Expr Expr 
          |Sub Expr Expr
          |Mul Expr Expr
          |Div Expr Expr
          |Expo Expr Expr
          |Cos Expr
          |Sin Expr
          deriving Show 

data Token = Tcon N 
           | TId Id
           | AddOp 
           | SubOp
           | MulOp 
           | DivOp
           | ExpOp 
           | CosOp
           | SinOp
           | OpAss
           | LPar | RPar
           | TExpr Expr 
           | TAss Id Expr 
           deriving Show 

--type of environment for variables to be assigned in
type Env = [(Id,N)]

--to find values that are associated with given variables
lookUp :: String -> Env -> Double
lookUp a [] = error "there is no pair associated with this String"
lookUp a ((x,y):zs) = if (a==x) then y else lookUp a zs

--this function adds a var to a env list 
addVar :: (String, Double) -> Env -> Env 
addVar x [] = [x]
addVar x env = x : env 

--when calling update we will need to call remove to replace the env 
removeVar :: String -> Env -> Env 
removeVar a [] = []
removeVar a ((x,y):zs) = if a == x then zs else (x,y) : removeVar a zs

--update will set value of given parameter to new value and keep everything else the same
update :: String -> Double -> Env -> Env 
update s d [] = [(s,d)]
update s d env = addVar (s,d) (removeVar s env)

eval :: Instr -> Env -> Either Env Double 
eval (Assign v e) env = Left (update (v) (exprEval env e) env) 
eval (AExpr e) env = Right ((exprEval env e))

--handling evaluating our exp 
exprEval ::Env -> Expr -> Double
exprEval env (Var x) = lookUp x env
exprEval env (Const x) = x
exprEval env (Add x xx) = (exprEval env x) + (exprEval env xx)
exprEval env (Sub x xx) = (exprEval env x) - (exprEval env xx)
exprEval env (Mul x xx) = (exprEval env x) * (exprEval env xx)
exprEval env (Div x xx) = (exprEval env x) / (exprEval env xx)
exprEval env (Expo x xx) = ((exprEval env x) ** (exprEval env xx))
exprEval env (Cos x ) = cos(exprEval env x) 
exprEval env (Sin x ) = sin(exprEval env x) 

------------------------- Now we are in preproccessing land here 
tokenize :: String -> [Token]
tokenize s = map classify $ words $ prep s

prep :: String -> String 
prep [] = []
prep ('(':xs) = " ( " ++ prep xs
prep (')':xs) = " ) " ++ prep xs
prep ('+':xs) = " + " ++ prep xs
prep ('-':xs) = " - " ++ prep xs
prep ('=':xs) = " = " ++ prep xs
prep ('*':xs) = " * " ++ prep xs
prep ('/':xs) = " / " ++ prep xs
prep ('^':xs) = " ^ " ++ prep xs
prep ('c':'o':'s':xs) = " cos " ++ prep xs
prep ('s':'i':'n':xs) = " sin " ++ prep xs
prep (x:xs) = x : prep xs

-- checking for correct number value 123 decimal followed by 12312 no more decimals 
isCSym:: String -> Bool 
isCSym [] = True
isCSym (x:xs) | x `elem` ['0' .. '9'] = isCSym xs 
              | x == '.' = isCSym' xs 
              | otherwise = False 
              where isCSym' [] = True
                    isCSym' (y:ys) = if y `elem` ['0' .. '9'] then isCSym' ys else False 

isVSym :: String -> Bool 
isVSym [] = False
isVSym (x:xs) =  if isLower x then isVSym' xs else False 

isVSym' :: String -> Bool
isVSym' [] = True
isVSym' (x:xs) = if (isLower x && isDigit x) then isVSym' xs else False

classify :: String -> Token
classify "(" = LPar
classify ")" = RPar
classify "+" = AddOp
classify "-" = SubOp
classify "=" = OpAss
classify "*" = MulOp
classify "/" = DivOp
classify "^" = ExpOp
classify "cos" = CosOp
classify "sin" = SinOp
classify x | isVSym x = TId x
           | isCSym x = Tcon (read x :: Double)
           | otherwise = error "classify error: Not a token"

 

-- data Token = Tcon N 
--            | TId Id
--            | AddOp 
--            | SubOp
--            | MulOp 
--            | DivOp
--            | ExpOp 
--            | CosOp
--            | SinOp
--            | OpAss
--            | LPar | RPar
--            | TExpr Expr 
--            | TAss Id Expr 
--            deriving Show 

-- here we are goin to do all the parsing in this file 
parse :: [Token] -> [Token] -> [Token]
parse (Tcon n : stack) input = parse (TExpr (Const n): stack) input

parse (RPar: TExpr e : LPar : stack)        input = parse (TExpr e : stack) input -- taking care of parens we just get rid of them 

parse (TExpr e2 : AddOp : TExpr e1 : stack) input = parse (TExpr (Add e1 e2) : stack) input -- covers add
parse (TExpr e2 : SubOp : TExpr e1: stack) input = parse (TExpr(Sub e1 e2) : stack) input -- covers Sub
parse (TExpr e2 : MulOp : TExpr e1 : stack) input = parse (TExpr (Mul e1 e2) : stack) input 
parse (TExpr e2 : DivOp : TExpr e1: stack) input = parse (TExpr(Div e1 e2) : stack) input 
parse (TExpr e2 : ExpOp : TExpr e1 : stack) input = parse (TExpr (Expo e1 e2) : stack) input 

parse (TExpr e : CosOp : stack)         input = parse (TExpr(Cos e):stack) input -- taking care of cos
parse (TExpr e : SinOp : stack)         input = parse (TExpr(Sin e):stack) input -- taking care of sin

parse (TExpr e2: OpAss : TId v : stack)     input = parse (TAss v e2 : stack ) input -- taking care of an assignmnet 

parse stack (i:input) = parse (i:stack) input
parse stack [] = stack


tokToIn :: [Token] -> Instr
tokToIn [] = error "Nothing to parse to instruction"
tokToIn [(TExpr e1)] = AExpr e1 
tokToIn [(TAss id e1)] = Assign id e1 
tokToIn _ = error "Parse Error"

tt :: String -> [Token]
tt s = tokenize s 

pp :: [Token] -> [Token]
pp s = parse [] s

tok :: [Token] -> Instr
tok s = tokToIn s

exec :: String -> Either Env Double
exec s = eval (tok (pp (tt s))) ([])